ai/anthropic/v0/main.go

```diff
@@ -0,0 +1,247 @@
+	Messages      []message   `json:"messages"`
+	MaxTokens     int         `json:"max_tokens"`
+	Metadata      interface{} `json:"metadata"`
+	StopSequences []string    `json:"stop_sequences,omitempty"`
+	Stream        bool        `json:"stream,omitempty"`
+	System        string      `json:"system,omitempty"`
+	Temperature   float32     `json:"temperature,omitempty"`
+	TopK          int         `json:"top_k,omitempty"`
+	TopP          float32     `json:"top_p,omitempty"`
+}
+
+type messagesOutput struct {
+	Text string `json:"text"`
+}
+
+type message struct {
+	Role    string    `json:"role"`
+	Content []content `json:"content"`
+}
+
+type usage struct {
+	InputTokens  int `json:"input_tokens"`
+	OutputTokens int `json:"output_tokens"`
+}
+
+// doesn't support anthropic tools at the moment
+type content struct {
+	Type   string  `json:"type"`
+	Text   string  `json:"text,omitempty"`
+	Source *source `json:"source,omitempty"`
+}
+
+type source struct {
+	Type      string `json:"type"`
+	MediaType string `json:"media_type"`
+	Data      string `json:"data"`
+}
+
+func Init(bc base.Component) *component {
+	once.Do(func() {
+		comp = &component{Component: bc}
+		err := comp.LoadDefinition(definitionJSON, setupJSON, tasksJSON, nil)
+		if err != nil {
+			panic(err)
+		}
+	})
+	return comp
+}
+
+type execution struct {
+	base.ComponentExecution
+	execute func(*structpb.Struct) (*structpb.Struct, error)
+	client  AnthropicClient
+}
+
+func (c *component) CreateExecution(sysVars map[string]any, setup *structpb.Struct, task string) (*base.ExecutionWrapper, error) {
+	e := &execution{
+		ComponentExecution: base.ComponentExecution{Component: c, SystemVariables: sysVars, Task: task, Setup: setup},
+		client:             newClient(getAPIKey(setup), getBasePath(setup), c.GetLogger()),
+	}
+	switch task {
+	case textGenerationTask:
+		e.execute = e.generateText
+	default:
+		return nil, fmt.Errorf("unsupported task")
+	}
+	return &base.ExecutionWrapper{Execution: e}, nil
+}
+
+func (e *execution) Execute(_ context.Context, inputs []*structpb.Struct) ([]*structpb.Struct, error) {
+	outputs := make([]*structpb.Struct, len(inputs))
+
+	// The execution takes a array of inputs and returns an array of outputs. The execution is done sequentially.
+	for i, input := range inputs {
+		output, err := e.execute(input)
+		if err != nil {
+			return nil, err
+		}
+
+		outputs[i] = output
+	}
+
+	return outputs, nil
+}
+
+func retriveMessageContent(contentsPbValue *structpb.ListValue, waitGroupLock *sync.WaitGroup, contents *[]content) {
+	defer waitGroupLock.Done()
+	for _, contentPbValue := range contentsPbValue.GetValues() {
+		contentType := contentPbValue.GetStructValue().Fields["type"].GetStringValue()
+		// anthrothpic models does not support image urls
+		if contentType == "text" {
+			content := content{
+				Type: "text",
+				Text: contentPbValue.GetStructValue().Fields["text"].GetStringValue(),
+			}
+			*contents = append(*contents, content)
```

Comment:
There is a high possibility of a thread-safety issue here. We should use a lock.
However, in this case, we don’t even need a goroutine since these operations are all CPU-bound. There’s no benefit to using a goroutine here.

Reply:
Ok, changed the implementation to 2 layer for loop.